packages/flutter/lib/src/widgets/scroll_delegate.dart

```diff
@@ -861,3 +862,239 @@ Widget _createErrorWidget(Object exception, StackTrace stackTrace) {
   FlutterError.reportError(details);
   return ErrorWidget.builder(details);
 }
+
+// TODO(Piinks): Come back and add keep alive support, <BUG>
+/// A delegate that supplies children for scrolling in two dimensions.
+///
+/// A [TwoDimensionalScrollView] lazily constructs its box children to avoid
+/// creating more children than are visible through the
+/// [TwoDimensionalViewport]. Rather than receiving children as an
+/// explicit [List], it receives its children using a
+/// [TwoDimensionalChildDelegate].
+///
+/// See also:
+///
+///   * [TwoDimensionalChildBuilderDelegate], an abstract subclass of this that
+///     lazily builds children on demand.
+///   * [TwoDimensionalChildListDelegate], an abstract subclass of this that
+///     uses a two dimensional array to layout children.
+abstract class TwoDimensionalChildDelegate extends ChangeNotifier {
+  /// Returns the child with the given [ChildVicinity], which is described in
+  /// terms of x and y indices.
+  ///
+  /// Subclasses must implement this function and will typically wrap their
+  /// children in [RepaintBoundary] widgets.
+  ///
+  /// The values returned by this method are cached. To indicate that the
+  /// widgets have changed, a new delegate must be provided, and the new
+  /// delegate's [shouldRebuild] method must return true.
+  Widget? build(BuildContext context, ChildVicinity vicinity);
+
+  /// Called whenever a new instance of the child delegate class is
+  /// provided.
+  ///
+  /// If the new instance represents different information than the old
+  /// instance, then the method should return true, otherwise it should return
+  /// false.
+  ///
+  /// If the method returns false, then the [build] call might be optimized
+  /// away.
+  bool shouldRebuild(covariant TwoDimensionalChildDelegate oldDelegate);
+}
+
+/// A delegate that supplies children for a [TwoDimensionalScrollView] using a
+/// builder callback.
+///
+/// The widgets returned from the builder callback are automatically wrapped in
+/// [RepaintBoundary] widgets if [addRepaintBoundaries] is true
+/// (also the default).
+///
+/// See also:
+///
+///  * [TwoDimensionalChildListDelegate], which is a similar delegate that has an
+///    explicit two dimensional array of children.
+abstract class TwoDimensionalChildBuilderDelegate extends TwoDimensionalChildDelegate {
+  /// Creates a delegate that supplies children for a [TwoDimensionalScrollView]
+  /// using the given builder callback.
+  ///
+  /// The [builder] and [addRepaintBoundaries] arguments must not be null.
+  TwoDimensionalChildBuilderDelegate({
+    this.addRepaintBoundaries = true,
+    required this.builder,
+    this.maxXIndex,
+    this.maxYIndex,
+  });
+
+  /// Called to build children on demand.
+  ///
+  /// Subclasses can choose to only call the builder for [ChildVicinity] indices
```

Comment:
So this is something I keep getting twisted on myself. The RenderTwoDimensionalViewport is what calls on the delegate to provide a child. Since RenderTwoDimensionalViewport is abstract, it is up to the subclass to decide what to do if/when null is returned for a ChildVicinity.

What I wanted to do here was leave the opportunity for developers to observe the common pattern of `null = end of children`, but also afford the ability for null to be valid, such as in the merged cell case for table.

Reply:
I think the choice you made makes sense. I would just add a little more color to the docs, e.g. that by default (= the default implementation of `build`) the builder is only called for indices between foo and bar, and subclasses can change this by overriding `build`. This just connects the dots a little more for people who want to do custom stuff.